app/models/discussion_reader.rb

```diff
@@ -25,23 +25,33 @@ def self.for_model(model, actor = nil)
     self.for(user: actor || model.author, discussion: model.discussion)
   end
 
-  def update_reader(read_at: nil, volume: nil, participate: false, dismiss: false)
-    viewed!(read_at, persist: false)    if read_at
+  def update_reader(ranges: nil, volume: nil, participate: false, dismiss: false)
+    viewed!(ranges, persist: false)     if ranges
     set_volume!(volume, persist: false) if volume && (volume != :loud || user.email_on_participation?)
     dismiss!(persist: false)            if dismiss
-    save(validate: false)               if changed?
+    save!                               if changed?
   end
 
-  def viewed!(read_at, persist: true)
-    return if self.last_read_at && self.last_read_at > read_at
-    assign_attributes(read_attributes(read_at))
-    EventBus.broadcast('discussion_reader_viewed!', self)
+  def viewed!(ranges = [], persist: true)
+    # byebug
+    # ranges = RangeSet.to_ranges(ranges)
```

Comment:
Do you want to keep this in to ensure you have a rangeset before proceeding? Good ol' ruby type check guarding, heh. I'm fine either way.

Reply:
na, we dont need it, which is nice